pkg/service/healthcheck/metrics.go

```diff
@@ -145,11 +143,19 @@ func apply(ch <-chan prometheus.Metric, f func(cluster, dc, host, pt string, v f
 	}
 }
 
-func collect(g *prometheus.GaugeVec) chan prometheus.Metric {
-	ch := make(chan prometheus.Metric, metricBufferSize)
+func collect(g *prometheus.GaugeVec) []prometheus.Metric {
+	ch := make(chan prometheus.Metric)
 	go func() {
 		g.Collect(ch)
 		close(ch)
 	}()
-	return ch
+
+	// Collect hold RLock during entire call. Read entire channel to prevent passing channel outside.
```

Comment:
nit: holds

Reply:
nit: Read the entire channel into an array to prevent passing the channel back where processing particular items could access the same lock and create a deadlock as the Collect function has nowhere to write to until the items are read.